### ISSUE TITLE
HPA: scaling replica count directly based on a metric

### ISSUE THREAD

**Issue by eero-t** [kubernetes/kubernetes#109920]
### What would you like to be added?

HPA policy where simpler algorithm is used: desiredReplicas = factor * currentMetricValue.

I'd like this to be supported at least for app/custom metrics.

### Why is this needed?

Current scaling algorithm: desiredReplicas = ceil[currentReplicas * ( currentMetricValue / desiredMetricValue )]

Has several issues:
* It's harder to understand (one can easily find several k8s bugs about its impact)
* It does not allow handling properly situations when there's a significant delay between deployment count change, and (custom) metric value change. In such situations it causes:
  * very excessive deployment counts on scale up (which is natural result of delay with currentReplicas*currentMetricValue relation),
  * massive amount of flapping (metric drops fast after the delay due to excessive deployment count, and the resulting sharp deployment count drop causes then immediately drastic metric increase)

One could avoid this by telling HPA to act slower than the delay, but when that delay can be significant, HPA can then be too slow to be useful.

These issues could be avoided by HPA supporting additional "desiredReplication = factor * currentMetricValue"  algorithm/policy.

Use-case example:
* Custom metric on service request queue (e.g. number of queued items not yet being processed)
* HPA that scales replica count for backends that serve these requests, based on that custom metric
* Requests taking tens of seconds to complete (= delay impacting custom metric change)
* Desired scaling response time being in single-digit seconds

**Comment by eero-t** [comment]
And https://github.com/kubernetes/kubernetes/issues/96242 will make the issue only worse.

**Comment by pbetkier** [comment]
I don't understand the problem. Isn't `AverageValue` target type exactly what you propose? ([docs](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#autoscaling-on-metrics-not-related-to-kubernetes-objects), [code](https://github.com/kubernetes/kubernetes/blob/b3be343bc8d7b74d1da44cf51df7a4bf73f63ed6/pkg/controller/podautoscaler/replica_calculator.go#L372))

It boils down to `replicas = ceil[currentMetricValue / target.averageValue]`, so it's the same as `replicas = factor * currentMetricValue` that you propose, where `target.averageValue` is `1 / factor` and ceil is applied for rounding.

In other words: if I want to have a pod for each 100 elements pending on a queue (target metric is queue length) then you can already configure `target.averageValue = 100`, whereas you propose introducing a new target type with `target.factor = 0.01`, which seems redundant.

Please clarify.

**Comment by eero-t** [comment]
@pbetkier Thanks!   Yes, it appears to provide what I need (I'll test it in few weeks).

I think this bug could be pivoted to be about documentation as this feature is appallingly documented:

* It's not on the main HPA page: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/
* On walkthrough page this important item is stuffed at the end of (a long) document, with a misleading title and also somewhat misleading description

=> No wonder me and others miss it

I think walkthrough would benefit a lot of just describing what it does, and removing misleading/extra fluff:

Title: "Autoscaling on metrics not related to Kubernetes objects" -> "Autoscaling directly on metrics"

(Firstly, the main point is scaling directly on metric values, not on Kubernetes object counts, so title should reflect that.  Secondly, of course all/most metrics used with this would be related to Kubernetes objects, just not directly.)

"Applications running on Kubernetes may need to autoscale based on metrics that don't have an obvious relationship to any object in the Kubernetes cluster, such as metrics describing a hosted service with no direct correlation to Kubernetes namespaces." -> 
"Applications running on Kubernetes may need to autoscale based on metrics that don't have direct relationship to any object in the Kubernetes cluster."

(relationship can be obvious although not direct, there's better example description later, and namespaces is irrelevant notice)

"Using external metrics requires knowledge of your monitoring system; the setup is similar to that required when using custom metrics." -> DROP (useless filler)

Description should include `replicas = ceil[currentMetricValue / target.averageValue]` as from preformatted text user can easily notice there being a different algorithm available while skimming the page content (or searching for `currentMetricValue`).

And obviously this alternative algorithm should be (at least) mentioned on the main HPA page too.

**Comment by pbetkier** [comment]
I agree the documentation is confusing! And thanks for the suggestions.

Also note https://github.com/kubernetes/kubernetes/issues/79272 which is also about making changes to HPA docs.
